app/code/core/Mage/Core/Model/Config.php

```diff
@@ -501,13 +491,44 @@ public function getCache()
     }
 
     /**
-     * Get lock flag cache identifier
+     * Call before building and saving cache to ensure only one process can save the cache
      *
-     * @return string
+     * If failed to get cache lock:
+     *   - CLI: throws exception
+     *   - Other: 503 error
+     *
+     * @return void
+     * @throws Exception
      */
-    protected function _getCacheLockId()
+    public function getCacheSaveLock()
     {
-        return $this->getCacheId() . '.lock';
+        if (!Mage::app()->useCache('config')) {
+            return;
+        }
+        $connection = Mage::getSingleton('core/resource')->getConnection('core_write');
+        $lockWaitTime = getenv('MAGE_CONFIG_CACHE_LOCK_WAIT') ?: (PHP_SAPI === 'cli' ? 60 : 3);
+        if (!$connection->fetchOne("SELECT GET_LOCK('core_config_cache_save_lock', ?)", [$lockWaitTime])) {
+            if (PHP_SAPI === 'cli') {
+                throw new Exception('Could not get lock on cache save operation.');
+            } else {
+                require_once Mage::getBaseDir() . DS . 'errors' . DS . '503.php';
+                die();
+            }
+        }
+    }
+
+    /**
+     * Release the cache saving lock after it is saved or no longer needed
+     *
+     * @return void
+     */
+    public function releaseCacheSaveLock()
+    {
+        if (!Mage::app()->useCache('config')) {
+            return;
+        }
```

Comment:
```suggestion
```

weird suggestion, I wouldn't release the lock only if cache is enabled, just in case between the "get lock" and the "release lock" something disables the cache and suddenly the lock is never released

does it make sense?

Reply:
There is no way for the `useCache` return value to change between the calls to `getCacheSaveLock` and `releaseCacheSaveLock` in the same process, so this is just added to avoid an unnecessary `SELECT RELEASE_LOCK(...` query rather than wrapping all of the calls to it with `if (..useCache..)` or using more complex if/else structures in the calling methods. Trying to favor readability but I guess that's subjective. The overhead of this method is extremely low after the first call, basically jsut checking local variables and an array by index.

Reply:
I was also thinking about wrapping everything in the "if usecache" check but at the end of the day your version is more readable and well written